components/fxa-client/src/oauth.rs

```diff
@@ -20,13 +20,27 @@ use std::{
     time::{SystemTime, UNIX_EPOCH},
 };
 use url::Url;
-
 // If a cached token has less than `OAUTH_MIN_TIME_LEFT` seconds left to live,
 // it will be considered already expired.
 const OAUTH_MIN_TIME_LEFT: u64 = 60;
+// If we see more than AUTH_CHECK_CIRCUIT_BREAKER_COUNT checks, and each is less than
+// AUTH_CHECK_CIRCUIT_BREAKER_RESET_MS since the last check, then we'll trigger a "circuit breaker".
+const AUTH_CHECK_CIRCUIT_BREAKER_RESET_MS: u64 = 1000 * 10;
+// This logic is in place to protect ourselves from endless auth recovery loops, while at the same
+// time allowing for a few 401s to hit the state machine in a quick succession.
+// For example, initializing the account state machine & syncing after letting our access tokens expire
+// due to long period of inactivity will trigger a few 401s, and that shouldn't be a cause for concern.
+const AUTH_CHECK_CIRCUIT_BREAKER_COUNT: u64 = 10;
 // Special redirect urn based on the OAuth native spec, signals that the
 // WebChannel flow is used
 pub const OAUTH_WEBCHANNEL_REDIRECT: &str = "urn:ietf:wg:oauth:2.0:oob:oauth-redirect-webchannel";
+
+#[derive(Clone, Copy)]
+pub(crate) struct AuthorizationCache {
```

Comment:
Naming nit: for me, the name "cache" doesn't create the right mental model; caching tends to imply some transient data that's kept around for performance reasons, but that's not what we're doing here.

I suggest leaning in to the terminology and calling it an `AuthCircuitBreaker` or similar. Moving some of the increment-and-check logic into a helper method might also aid future readers in understanding what's happening here. Strawman suggestion:

```
struct AuthCircuitBreaker { ... }

impl AuthCircuitBreaker {
  fn check(&self) -> Result<()> {
    // Increment `self.attempts`, reset the timestamp, etc.
    // Return an error if `self.attempts` exceeds the threshold.
  }
}
```

And then in your `check_authorization_status` method:

```
impl FirefoxAccount {
  fn check_authorization_status(&mut self) -> Result<IntrospectInfo> {
    self.auth_circuit_breaker.check()?;
    // The rest of the method body, as before.
  }
}
```